File: binarysearch.c
#include <stdio.h>

int binarysearch(int * arr, int item, int arrsize){
    int low = 0;
    int high = arrsize - 1;

    while(low <= high){
        int mid = (low+high)/2;
        int guess = arr[mid];
        if(guess == item){
            return mid;
        } else if(guess > item){
            high = mid -1;
        } else {
            low = mid + 1;
        }
    }
    return -1;
}

int main(){
    int arr[] = {1, 3, 5, 7, 9};
    int arrsize = sizeof(arr)/sizeof(arr[0]);
    int item = 3;
    int result = binarysearch(arr, item, arrsize);
    if(result == -1){
        printf("Item not found\n");
    } else {
        printf("Item found at index %d\n", result);
    }
    return 0;
}

File: binarysearch.py
import time
from functools import wraps


def runtime(func):
    @wraps(func)
    def wrapper(arr: list, item):
        start = time.time()
        res = func(arr, item)
        end = time.time()
        print(f'program run in {end-start} seconds')
        return res
    return wrapper


# @runtime
def binsearch(arr: list, item):
    print(arr)
    low = 0
    n = len(arr) - 1
    mid = (low + n)//2
    if arr[mid] == item:
        return mid
    else:
        if arr[mid] > item:
            return binsearch(arr[:mid], item)
        elif arr[mid] < item:
            res = binsearch(arr[mid+1:], item)
            return mid + 1 + res

start = time.time()
print('(binarysearch 001): element found at index: ',
      binsearch([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 4))
end = time.time()
print(f"program run in {end-start} seconds")


@runtime
def binsrch(arr: list, item):
    def _binsrch(arr: list, item, low, high):
        print(arr)
        mid = (high + low) // 2
        if arr[mid] == item:
            return mid
        elif arr[mid] > item:
            return _binsrch(arr, item, low, mid - 1)
        else:
            return _binsrch(arr, item, mid + 1, high)
    return _binsrch(arr, item, 0, len(arr) - 1)


print('(binarysearch 002): element found at index: ',
      binsrch([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 4))
